framework/src/media/audio/audio_manager.c

```diff
@@ -1299,25 +1350,281 @@ audio_manager_result_t find_stream_in_device_with_process_type(device_process_ty
+	if (fd < 0) {
+		meddbg("Failed to open audio driver, path : %s errno : %d\n", path, errno);
+		pthread_mutex_unlock(&(card->card_mutex));
+		return AUDIO_MANAGER_NO_AVAIL_CARD;
+	}
+	process_type = get_process_type_audio_param_value(type);
+
+	subprocess_type = get_subprocess_type_audio_param_value(subtype);
+	cap_desc.caps.ac_len = sizeof(struct audio_caps_s);
+	cap_desc.caps.ac_type = process_type;
+	cap_desc.caps.ac_subtype = subprocess_type;
+
+	ret = ioctl(fd, AUDIOIOC_CONFIGURE, (unsigned long)&cap_desc);
+	if (ret < 0) {
+		if (ret == -EINVAL) {
+			meddbg("Device doesn't support it!\n");
+			ret = AUDIO_MANAGER_DEVICE_NOT_SUPPORT;
+		} else {
+			meddbg("ioctl failed ret : %d\n", ret);
+			ret = AUDIO_MANAGER_OPERATION_FAIL;
+		}
+		close(fd);
+		pthread_mutex_unlock(&(card->card_mutex));
+		return ret;
+	}
+	close(fd);
+	pthread_mutex_unlock(&(card->card_mutex));
+	return AUDIO_MANAGER_SUCCESS;
+}
+
+audio_manager_result_t register_stream_in_device_process_handler(int card_id, int device_id, device_process_type_t type)
+{
+	char path[AUDIO_DEVICE_FULL_PATH_LENGTH];
+	char mq_path[AUDIO_DEVICE_PROCESS_QUEUE_PATH_LENGTH];
+	int fd;
+	audio_card_info_t *card;
+	audio_config_t *config;
+	uint8_t process_type;
+	int ret;
+	struct mq_attr attr;
+
+	card = &g_audio_in_cards[card_id];
+	config = &card->config[device_id];
+	get_card_path(path, card_id, device_id, INPUT);
+
+	pthread_mutex_lock(&(card->card_mutex));
+	fd = open(path, O_RDONLY);
+	if (fd < 0) {
+		meddbg("Failed to open audio driver, path : %s errno : %d\n", path, errno);
+		pthread_mutex_unlock(&(card->card_mutex));
+		return AUDIO_MANAGER_NO_AVAIL_CARD;
+	}
+	process_type = get_process_type_audio_param_value(type);
+
+	pthread_mutex_unlock(&(card->card_mutex));
+
+	if ((get_supported_process_type(fd, config, INPUT) != AUDIO_MANAGER_SUCCESS) || (((config->device_process_type) & process_type) == 0)) {
+		meddbg("Device doesn't support it\n");
+		close(fd);
+		return AUDIO_MANAGER_DEVICE_NOT_SUPPORT;
+	}
+
+	pthread_mutex_lock(&(card->card_mutex));
+
+	/* Create a message queue */
+	attr.mq_maxmsg = 16;
+	attr.mq_msgsize = sizeof(struct audio_msg_s);
+	attr.mq_curmsgs = 0;
+	attr.mq_flags = 0;
+	snprintf(mq_path, AUDIO_DEVICE_PROCESS_QUEUE_PATH_LENGTH, "%s%s", path, AUDIO_DEVOCE_PROCESS_QUEUE_PATH);
+	config->process_handler = mq_open(path, O_RDWR | O_CREAT, 0644, &attr);
+
+	/* Register our message queue with the audio device */
+	ret = ioctl(fd, AUDIOIOC_REGISTERPROCESS, (unsigned long)config->process_handler);
+	if (ret < 0) {
+		if (ret == -EBUSY) {
+			meddbg("Already Registered\n");
+			ret = AUDIO_MANAGER_DEVICE_ALREADY_IN_USE;
+		} else if (ret == -EINVAL) {
+			meddbg("Device doesn't support it!\n");
+			ret = AUDIO_MANAGER_DEVICE_NOT_SUPPORT;
+		} else {
+			meddbg("Register Handler Error : %d\n", ret);
+			ret = AUDIO_MANAGER_DEVICE_FAIL;
+		}
+		close(fd);
+		pthread_mutex_unlock(&(card->card_mutex));
+		return ret;
+	}
+	pthread_mutex_unlock(&(card->card_mutex));
```

Comment:
Doesn't fd(close) need to be added before this pthread_mutex_unlock?

Reply:
fixed.